### PR TITLE
Long callback refactor

### LINKED ISSUE #1769: [BUG] long_callback short interval makes job run in loop
I'm really excited to use `long_callback`s because it helps a lot for my use-cases, but [while reading the doc](https://dash.plotly.com/long-callbacks) (Example 5) I found that there exists an undocumented parameter `interval` which is set to `1000` in this example.

I tried setting it to `1000` too, but the progress updates were too slow, so I progressively lowered it. It behaved as I wanted when setting it to `100` (ms), but when running my job with `interval=100`, when the job finishes, it is immediatly executed again, even though the launch button's event is not triggered.

## Minimal replication code

```python3
#!/usr/bin/env python3
import time
import dash
import dash_html_components as html
import dash_core_components as dcc
from dash.long_callback import DiskcacheLongCallbackManager
from dash.dependencies import Input, Output
import plotly.graph_objects as go

## Diskcache
import diskcache
cache = diskcache.Cache("./cache")
long_callback_manager = DiskcacheLongCallbackManager(cache)

def make_progress_graph(progress, total):
    progress_graph = (
        go.Figure(data=[go.Bar(x=[progress])])
        .update_xaxes(range=[0, total])
        .update_yaxes(
            showticklabels=False,
        )
        .update_layout(height=100, margin=dict(t=20, b=40))
    )

    return progress_graph


app = dash.Dash(__name__, long_callback_manager=long_callback_manager)

app.layout = html.Div(
    [
        html.Div(
            [
                html.P(id="paragraph_id", children=["Button not clicked"]),
                dcc.Graph(id="progress_bar_graph", figure=make_progress_graph(0, 10)),
            ]
        ),
        html.Button(id="button_id", children="Run Job!"),
        html.Button(id="cancel_button_id", children="Cancel Running Job!"),
        html.Span(id="heavy-data-span", children="", style={"display": "none"}),
        html.Span(id="output-span", children="", style={"display": "none"}),
    ]
)

@app.long_callback(
    output=[Output("paragraph_id", "children"),Output("output-span", "children"),],
    inputs=Input("button_id", "n_clicks"),
    running=[
        (Output("button_id", "disabled"), True, False),
        (Output("cancel_button_id", "disabled"), False, True),
        (
            Output("paragraph_id", "style"),
            {"visibility": "hidden"},
            {"visibility": "visible"},
        ),
        (
            Output("progress_bar_graph", "style"),
            {"visibility": "visible"},
            {"visibility": "hidden"},
        ),
    ],
    cancel=[Input("cancel_button_id", "n_clicks")],
    progress=[Output("progress_bar_graph", "figure"), Output("heavy-data-span", "children")],
    progress_default=(make_progress_graph(0, 20), "A"),
    interval=38,
    prevent_initial_call=True,
)
def callback(set_progress, n_clicks):
    total = 20

    for i in range(total):
        print(f"Running {i}/{total}...")
        time.sleep(0.3)
        set_progress((make_progress_graph(i, total), str(i%10)*4_000_000))

    return [f"Clicked {n_clicks} times", "A"*4_000_000]


if __name__ == "__main__":
    app.run_server(debug=True)

```
If you can't replicate this exact looping bug on your setup, try reducing the `interval` value until you see a looping behaviour with some `diskcache` I/O errors, and when you do, progressively set it to something higher until the looping remains, but that the error disappear.

Here is what such diskcache I/O errors look like, though in my real application I never witnessed these. There simply were no warnings !
```
Traceback (most recent call last):
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/multiprocess/process.py", line 315, in _bootstrap
    self.run()
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/multiprocess/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/dash/long_callback/managers/diskcache_manager.py", line 143, in job_fn
    user_callback_output = fn(*maybe_progress, *user_callback_args)
  File "/Users/sam1902/Downloads/test_dash_bug.py", line 78, in callback
    set_progress(make_progress_graph(i, total))
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/dash/long_callback/managers/diskcache_manager.py", line 137, in _set_progress
    cache.set(progress_key, progress_value)
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/diskcache/core.py", line 799, in set
    with self._transact(retry, filename) as (sql, cleanup):
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/contextlib.py", line 117, in __enter__
 58         ),
    return next(self.gen)
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/diskcache/core.py", line 713, in _transact
    sql = self._sql
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/diskcache/core.py", line 651, in _sql
    return self._con.execute
  File "/Users/sam1902/.pyenv/versions/3.9.1/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/diskcache/core.py", line 626, in _con
    con = self._local.con = sqlite3.connect(
sqlite3.OperationalError: disk I/O error
```
Those are not critical errors causing the server to crash. It simply ends one of the processes, but the callback still loops back. Also, those happen anywhere within the loop (at any iteration, not just at the end).
## Environment
`pip list | grep dash`:
```
dash                      2.0.0
dash-bootstrap-components 0.13.0
dash-core-components      2.0.0
dash-html-components      2.0.0
dash-table                5.0.0
diskcache                  5.2.1
psutil                     5.8.0
multiprocess               0.70.12.2
```
-  if frontend related, tell us your Browser, Version and OS

    - OS: macOs 11.5.2
    - Browser: replicated on
      - Firefox 92.0 (i.e. latest on 2021-09-21)
      - Chromium 93.0.4577.82 (i.e. latest on 2021-09-21)

**Describe the bug**

When setting an `interval` too small, the callback runs, but once it finishes it runs again, and again etc in a loop fashion.
From my replication attempts, this seems to be due to passing too large values to the long_callback's Output or `progress=[Output(..)]`, either during progress updates or at the end of the callback.

I ran into this in my code because I dealt with large images encoded as data URIs, and the file size here are approximated by `"A"*4_000_000`, which should weigh around 3.8MB, which is comparable to my image weight.

**Expected behavior**

The callback should only be called once, no matter the progress callback frequency (i.e. interval time)

**Screenshots**
As you can see, it loops from `Runnning 19/20` back to `Running 0/20` **without any warning or error**:
Screenshot 2021-09-21 at 11 52 25

### LINKED ISSUE #1821: [BUG] Long Callbacks: Raising Error Causes Infinite Loop (Including PreventUpdate)
**Environment**

```
python               3.9.7
dash                 2.0.0
```

**Describe the bug**

If an error is raised within a long callback, an infinite loop of re-raising this error is triggered, as a new process tries again and again
This is triggered by the `_long_callback_interval_1.n_intervals` property

**Expected behavior**

The error would be handled like with normal callbacks, and the long callback wouldn't be called by new processes

**Minimum Reproducible Example**

```
import time

import dash
import diskcache
from dash import html
from dash.dependencies import Input, Output
from dash.long_callback import DiskcacheLongCallbackManager

cache = diskcache.Cache("./cache")
long_callback_manager = DiskcacheLongCallbackManager(cache)

app = dash.Dash(__name__, long_callback_manager=long_callback_manager)

app.layout = html.Div(
    [
        html.Div([html.P(id="paragraph_id", children=["Button not clicked"])]),
        html.Button(id="button_id", children="Run Job!"),
    ]
)


@app.long_callback(
    output=Output("paragraph_id", "children"),
    inputs=Input("button_id", "n_clicks"),
    running=[
        (Output("button_id", "disabled"), True, False),
    ],
)
def callback(n_clicks):
    if n_clicks == 2:
        raise dash.exceptions.PreventUpdate(dash.callback_context.triggered[0]['prop_id'])
    time.sleep(2.0)
    return [f"Clicked {n_clicks} times"]


if __name__ == "__main__":
    app.run_server(debug=True)
```

In the above example, can safely click the button once, then on the second click raising `PreventUpdate` causes the infinite loop
With logged messages of:

```
Process Process-3:
Traceback (most recent call last):
  File "./python3.9/site-packages/multiprocess/process.py", line 315, in _bootstrap
    self.run()
  File "./python3.9/site-packages/multiprocess/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "./python3.9/site-packages/dash/long_callback/managers/diskcache_manager.py", line 143, in job_fn
    user_callback_output = fn(*maybe_progress, *user_callback_args)
  File "./reprex_long_callback_infinite_loop_on_error.py", line 31, in callback
    raise dash.exceptions.PreventUpdate(dash.callback_context.triggered[0]['prop_id'])
dash.exceptions.PreventUpdate: button_id.n_clicks
Process Process-4:
Traceback (most recent call last):
  File "./python3.9/site-packages/multiprocess/process.py", line 315, in _bootstrap
    self.run()
  File "./python3.9/site-packages/multiprocess/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "./python3.9/site-packages/dash/long_callback/managers/diskcache_manager.py", line 143, in job_fn
    user_callback_output = fn(*maybe_progress, *user_callback_args)
  File "./reprex_long_callback_infinite_loop_on_error.py", line 31, in callback
    raise dash.exceptions.PreventUpdate(dash.callback_context.triggered[0]['prop_id'])
dash.exceptions.PreventUpdate: _long_callback_interval_1.n_intervals
Process Process-5:
Traceback (most recent call last):
  File "./python3.9/site-packages/multiprocess/process.py", line 315, in _bootstrap
    self.run()
  File "./python3.9/site-packages/multiprocess/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "./python3.9/site-packages/dash/long_callback/managers/diskcache_manager.py", line 143, in job_fn
    user_callback_output = fn(*maybe_progress, *user_callback_args)
  File "./reprex_long_callback_infinite_loop_on_error.py", line 31, in callback
    raise dash.exceptions.PreventUpdate(dash.callback_context.triggered[0]['prop_id'])
dash.exceptions.PreventUpdate: _long_callback_interval_1.n_intervals
Process Process-6:
Traceback (most recent call last):
  File "./python3.9/site-packages/multiprocess/process.py", line 315, in _bootstrap
    self.run()
  File "./python3.9/site-packages/multiprocess/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "./python3.9/site-packages/dash/long_callback/managers/diskcache_manager.py", line 143, in job_fn
    user_callback_output = fn(*maybe_progress, *user_callback_args)
  File "./reprex_long_callback_infinite_loop_on_error.py", line 31, in callback
    raise dash.exceptions.PreventUpdate(dash.callback_context.triggered[0]['prop_id'])
dash.exceptions.PreventUpdate: _long_callback_interval_1.n_intervals
Process Process-7:
...
```

### LINKED ISSUE #1852: [BUG] Long Callback interval_time causing infinite loop
```
dash 2.0.0
python 3.8.8
diskcache 5.3
waitress 2.0
```
    - OS: Windows 10
    - Browser: Firefox 94.0.2, Chrome 96.0.4, Edge 96.0

**The bug**

Implemented a long callback for a long processing function that takes 60 - 120 seconds. The browser will start checking if the callback is finished once per second. If I run the dash code on my PC on http://127.0.0.1:8080 using either the built in Flask server or Waitress, the page will load without an issue. When deployed to a server, it will enter into an infinite loop. After 60 seconds, the title of the website will turn to "Updating..." and the content will not load. When using the built in network monitor in Firefox, there are POST requests submitted by the browser every second, followed by the the browser attempting to download data and stopping at the next post request. Due to the speed of the download, the POST request never completes and the page doesn't load.

By changing the hard coded interval_time in dash.py from 1000 to 10000 I was able to eliminate this issue. With the new settings, Firefox network monitor timed the POST response to be 3.5 seconds.

```
    def long_callback(self, *_args, **_kwargs):
        ...
        interval_time = _kwargs.pop("interval", 10000)

```

**Expected behavior**

When the long callback is ready for the user's browser to download, the interval_component should be disabled in an initial POST, followed by a second POST with the full data set. In the current behavior, a single POST disables the interval and delivers the full data set. If the download time exceeds the interval time, then the browser will not receive the data.

### LINKED ISSUE #1974: [BUG] Long Callbacks: dash.no_update not supported with Celery/Redis implementations
**Environment**

```
python               3.9.7
dash                 2.3.0
dash-core-components 2.0.0
dash-html-components 2.0.0
dash-table           5.0.0
```

**Describe the bug**

Cannot return `dash.no_update` from a celery/redis implementation of Long Callbacks. Raises `TypeError: Object of type NoUpdate is not JSON serializable` (seen in celery logs). 

**Expected behavior**

Can return `dash.no_update` as an Output, as in DiskLongCallbackManager.

**Minimum Reproducible Example**

```
import dash
from celery import Celery
from dash import html, no_update
from dash.dependencies import Input, Output
from dash.long_callback import CeleryLongCallbackManager

celery_app = Celery(broker='redis://0.0.0.0:6379/0', backend='redis://0.0.0.0:6379/0')
long_callback_manager = CeleryLongCallbackManager(celery_app)

application = dash.Dash(__name__, long_callback_manager=long_callback_manager)

application.layout = html.Div(
    [
        html.Div([html.P(id="paragraph_id", children=["Callback has not updated"])]),
        html.Button(id="button_id", children="Run Job!"),
        html.Div([html.P(id="no-update", children=["Testing_no_update"])]),
    ]
)


@application.long_callback(
    output=[Output("paragraph_id", "children"),
                   Output("no-update", "children")],
    inputs=Input("button_id", "n_clicks")
)
def callback(n_clicks):
    return [f"Clicked {n_clicks} times"], no_update


if __name__ == "__main__":
    application.run_server(debug=True)
```

In the above example we expect the app to render `Clicked None times` on initial load, as with DiskLongCallbackManager. Instead we see `Callback has not updated` i.e. the long callback has not ran.

Logged messages:
```
Traceback (most recent call last):
  File "celery/app/trace.py", line 451, in trace_task
    R = retval = fun(*args, **kwargs)
  File "celery/app/trace.py", line 734, in __protected_call__
    return self.run(*args, **kwargs)
  File "dash/long_callback/managers/celery_manager.py", line 134, in job_fn
    cache.set(result_key, json.dumps(user_callback_output, cls=PlotlyJSONEncoder))
  File "json/__init__.py", line 234, in dumps
    return cls(
  File "_plotly_utils/utils.py", line 59, in encode
    encoded_o = super(PlotlyJSONEncoder, self).encode(o)
  File "json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "_plotly_utils/utils.py", line 136, in default
    return _json.JSONEncoder.default(self, obj)
  File "json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type NoUpdate is not JSON serializable
[2022-03-15 16:41:23,112: INFO/MainProcess] Task long_callback_9c77f4d8f656d9655d05d7238e789c68040815b5[601c3d29-60a5-4dfe-8925-16da1ab020cc] received
[2022-03-15 16:41:23,116: ERROR/ForkPoolWorker-8] Task long_callback_9c77f4d8f656d9655d05d7238e789c68040815b5[601c3d29-60a5-4dfe-8925-16da1ab020cc] raised unexpected: TypeError('Object of type NoUpdate is not JSON serializable')
```

### PR DESCRIPTION
**Pull request by T4rk1n**
Fix raising errors in long callback triggering an infinite update loop. Fix #1821
- raise PreventUpdate output a `no_content`
- Other errors are printed to console and appears in dev tools ui errors.
- Handle NoUpdate in celery long callbacks. Fix #1974

Replace long callback interval polling
- Fix short interval loop by handling the status of the long callback internally in the renderer when sending callback requests if not completed if it will refetch after `interval` milliseconds.
  - Fix #1769
  - Fix #1852 
- Add `long=False` to `dash.callback` to use instead of `app.long_callback`.
   - Prefix previous `app.long_callback` arguments with `long_` (`interval`, `running`, `cancel`, `progress`, `progress_default`, `cache_args_to_ignore`, `manager`)

Long callback parity:
- Support pattern matching.
- Support callback context.

### CODE REVIEW

File: dash/dash.py [hunk 1]
```diff
@@ -1228,6 +1240,28 @@ def callback(_triggers, user_store_data, user_callback_args):
...
+                            user_callback_output=NoUpdate(),
+                            in_progress=[val for (_, _, val) in running],
+                            interval_disabled=pending_job is None,
+                            progress=clear_progress,
+                            user_store_data=user_store_data,
+                        )
+
+                    if (
+                        isinstance(result, dict)
+                        and result.get("no_update") == "no_update"
```
Comment (alexcjohnson) [inline dash/dash.py:1261]: I guess having `no_update` create a value that's invalid as JSON would be tricky, particularly in the multi-output case but let's make the no_update serialization a little harder to encounter accidentally. I think we should (a) use the `_dash_` prefix we use for other internal-use values, ie `{"_dash_no_update": "_dash_no_update"}` , and (b) test for an exact match, not just this one key.

And re: multi-outputs, does it work to use `no_update` as one of several outputs to a long-callback? Can we include a test of this case?

**Comment by alexcjohnson** [comment]
Now that long callbacks are their own category to the renderer rather than creating more regular callbacks, a couple of extra questions:
- What happens if one of the extra outputs (progress or running) is an input to another callback? I guess that callback should fire every time that input changes, even if that's many times during one execution of the long callback, including after it completes. Does it?
- Do those extra outputs appear in the callback graph viz in the devtools? I'm not sure what this _should_ look like, probably not the same as the final outputs of the long callback but they should be included somehow. Not a super high priority though, we can come back to that if it's more than a quick adjustment.
- What about PMC inputs & outputs? I think you mentioned you expect these are now supported, but there are a number of situations to consider (and test):
  - `MATCH` - can we make a list of multiple independent input/output sets, each triggering its own instance of the long callback?
  - And does that include the extra inputs and outputs, like can you use `MATCH` to associate one progress bar and one cancel button with each long callback?
  - `ALL` - what happens if you trigger a long callback, then add or remove an item in the set? Does the running callback get canceled automatically?

File: dash/_callback.py [hunk 2]
```diff
@@ -17,22 +25,47 @@
...
+    long_interval=1000,
+    long_progress=None,
+    long_progress_default=None,
+    long_running=None,
+    long_cancel=None,
+    long_manager=None,
+    long_cache_args_to_ignore=None,
+    config_prevent_initial_callbacks=False,
+    callback_map=None,
+    callback_list=None,
```
[the remark this answers is not in the document]
Comment (T4rk1n) [reply dash/_callback.py:64]: > Am I understanding right that config_prevent_initial_callbacks, callback_map, and callback_list are not intended to be used by app developers, only by us internally? If that's the case, can we maybe hide them inside _kwargs so they don't become implicitly part of the public API?

Yes, good point.

**Comment by T4rk1n** [comment]
@alexcjohnson 

> What happens if one of the extra outputs (progress or running) is an input to another callback? I guess that callback should fire every time that input changes, even if that's many times during one execution of the long callback, including after it completes. Does it?

Tried an input on the progress, doesn't update, I used `dispatch(updateProps(...` for the side updates, what action should it use?

> Do those extra outputs appear in the callback graph viz in the devtools? 

No.

**Comment by T4rk1n** [comment]
> What about PMC inputs & outputs? 

Pattern matching callbacks are now supported for regular inputs/outputs.

> And does that include the extra inputs and outputs, like can you use MATCH to associate one progress bar and one cancel button with each long callback?

No, progress/running/cancel pattern matching are not supported and will throw an error.

> ALL - what happens if you trigger a long callback, then add or remove an item in the set? Does the running callback get canceled automatically?

No, the long callback continues until canceled/completed, not sure how that would be done.

**Comment by alexcjohnson** [comment]
> > Do those extra outputs appear in the callback graph viz in the devtools?
>
> No.

OK. It's going to be a little bit tricky to figure out what that should look like so let's make an issue to come back to this and any other limitations we think could potentially be removed later.

> > What about PMC inputs & outputs?
> 
> Pattern matching callbacks are now supported for regular inputs/outputs.

🌟 

> > And does that include the extra inputs and outputs, like can you use MATCH to associate one progress bar and one cancel button with each long callback?
> 
> No, progress/running/cancel pattern matching are not supported and will throw an error.

OK - as long as it's a clear error we can leave that for later as well.

> > ALL - what happens if you trigger a long callback, then add or remove an item in the set? Does the running callback get canceled automatically?
> 
> No, the long callback continues until canceled/completed, not sure how that would be done.

I guess for now I don't really care whether the long callback is allowed to continue (although I'd consider that another limitation, at some point we should make it be canceled as soon as it's known to be moot), but it's important that the first result is ignored when it finally is returned, if at that point there's a different set of inputs.

Here's a sample app that shows the desired behavior: if I type something, then change the number of inputs before the output callback has returned, I should not see the intermediate result appear, only the final state. That's how it works right now with regular callbacks, we should ensure it works the same with long callbacks.

```py
from dash import Input, Output, State, Dash, html, dcc, callback, ALL
import time


app = Dash(__name__)

app.layout = html.Div(
    [
        dcc.Dropdown(id="dropdown", value=1, options=list(range(10))),
        html.Div([], id="container"),
        html.Div(id="out")
    ],
)

@callback(Output("container", "children"), Input("dropdown", "value"), State("container", "children"))
def make_inputs(n, prev_children):
    prev_len = len(prev_children)
    if n > prev_len:
        return prev_children + [dcc.Input(id={"n": n}, value="") for n in range(prev_len, n)]
    return prev_children[:n]


@callback(Output("out", "children"), Input({"n": ALL}, "value"))
def out(vals):
    time.sleep(3)
    return "+".join(vals)


if __name__ == "__main__":
    app.run_server(debug=True)
```

**Comment by T4rk1n** [comment]
> I guess for now I don't really care whether the long callback is allowed to continue (although I'd consider that another limitation, at some point we should make it be canceled as soon as it's known to be moot), but it's important that the first result is ignored when it finally is returned, if at that point there's a different set of inputs.

It looked like the previous output was discarded in the case of regular callbacks (but job still running). For long pattern matching inputs, the output were not discarded. I've put a check on the output to auto cancel previous jobs.

**Review by alexcjohnson — APPROVED** [review]
💃 Looks great! Once the tests pass, the only other thing I'd like to see before merging is an issue listing the remaining known TODO items:
- pattern-matching with progress / cancel
- progress/cancel in devtools callback graph viz
- do all obsolete jobs now get canceled ASAP, or just the PMC jobs?
- anything else?

**Comment by T4rk1n** [comment]
> do all obsolete jobs now get canceled ASAP, or just the PMC jobs?

Yes

I'll create issues for devtools and pattern matching for progress/cancel/running.

**Pull request merged**